### PR TITLE
[MBUILDCACHE-86] bugfix / enhancements restoration of outputs on disk

### PR DESCRIPTION
**Pull request by kbuntrock**
Fix https://issues.apache.org/jira/browse/MBUILDCACHE-86

This PR covers some bugfix / enhancements with the restoration of outputs on disk.

It is in draft since I expect some discussions. And the IT test has still do be coded.

"Au menu" :
 
- Bugfix / "todo" : files in a base directory containing an underscore were wrongly restored to disk (not at the same location).
  -> To do so, the path is not guessed anymore from the classifier. I introduced a "filePath" property in the "attachedArtifact" section of the buildinfo.xml file.
  -> Because the buildInfo structure change, I changed the cache implementation version from "v1" to "v1.1". I assume it was one of the purpose of this value : we don't have to deal with structure migration. Any previous cache entry is defacto invalidated.
- Forbid the possibility to extract/restore data in a directory outside the project (like extracting `../../../.ssh` for example)
  -> I guess the extraction part is not a vulnerability since someone with commit permissions can guess other ways to extract data. But the possibility of restoring at any place on the disk looks pretty dangerous to me if a remote cache server is compromised.
- Gives the possibility to restore artefacts on disk, with a dedicated property : `maven.build.cache.restoreOnDiskArtefacts` (default to true, open for discussion)
- Introduce "globs" to filter extra attached outputs by filenames.

Following this checklist to help us incorporate your
contribution quickly and easily:

- [x] Make sure there is a [MBUILDCACHE JIRA issue](https://issues.apache.org/jira/browse/MBUILDCACHE) filed
for the change (usually before you start working on it). Trivial changes like typos do not
require a JIRA issue. Your pull request should address just this issue, without
pulling in other changes.
- [x] Each commit in the pull request should have a meaningful subject line and body.
- [x] Format the pull request title like [MBUILDCACHE-XXX] - Fixes bug in ApproximateQuantiles,
where you replace MBUILDCACHE-XXX with the appropriate JIRA issue. Best practice
is to use the JIRA issue title in the pull request title and in the first line of the
commit message.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Run mvn clean verify to make sure basic checks pass. A more thorough check will
be performed on your pull request automatically.
- [x] You have run the [Core IT](https://maven.apache.org/core-its/core-it-suite/) successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

- [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
- [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by jlstephens89** [comment]
@kbuntrock My company ran into the issue with clean deleting out the target directories but then not restoring the Jars when the build happens. I tested out your changes against our project and can confirm the jar restoration works as expected. Thank you for adding this! 🎉

**Comment by kbuntrock** [comment]
> @kbuntrock My company ran into the issue with clean deleting out the target directories but then not restoring the Jars when the build happens. I tested out your changes against our project and can confirm the jar restoration works as expected. Thank you for adding this! 🎉

Thank you. :) I guess it's more related to issue #92 

Related to this one, since there is less discussion than expected, I will code some tests next weekend and then fully submit the PR. :)

File: src/main/java/org/apache/maven/buildcache/artifact/OutputType.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.buildcache.artifact;
+
+public enum OutputType {
+    // generated project artefact
+    ARTEFACT(""),
```
Comment (AlexanderAshitkin) [inline src/main/java/org/apache/maven/buildcache/artifact/OutputType.java:23]: typo "I"

**Comment by martincekodhima** [comment]
Hey, we are running into the same issue as https://github.com/apache/maven-build-cache-extension/pull/104#issuecomment-1789473298. If the build is cached, the target folder is not being generated and neither is the jar. This causes issues if you have a script that depends on the jar being there. 

@kbuntrock Was that fixed in https://github.com/apache/maven-build-cache-extension/pull/92 or is that fixed with this MR?

**Comment by nobesio** [comment]
Hi folks! We are experiencing the same issue that @j-s-3 shared here:
> @kbuntrock My company ran into the issue with clean deleting out the target directories but then not restoring the Jars when the build happens. I tested out your changes against our project and can confirm the jar restoration works as expected. Thank you for adding this! 🎉

We were wondering if this PR will address it. Thank you!

**Comment by j-s-3** [comment]
@martincekodhima @nobesio we've been running a build from this branch since November and it has fixed that issue for us!

**Comment by nobesio** [comment]
Thank you for confirming this, @j-s-3! 

@kbuntrock & @AlexanderAshitkin: Is there anything we could do to help you merge and release these changes? I see there is just a comment related a typo.

**Comment by kbuntrock** [comment]
Since no one raised any concerns about to this PR (and even the opposite), I guess I can continue to work on the next steps.
I guess the key element is time. ;)

I have to check how I left this PR in term of tests.

For information, I use on my projects since October a custom version of the cache extension. Including slight modification of this PR. I have to check what was the modification about.

**Comment by olamy** [comment]
@kbuntrock Thanks for looking after this. I'm planning to release the cache extension in the coming weeks.
Not sure when do you plan to finish this PR? Would you mind creating a jira for it?
Thanks again

**Comment by hboutemy** [comment]
@kbuntrock I see you created MBUILDCACHE-86 and renamed the PR title
can you also update the Git commit to `[MBUILDCACHE-86]` instead of `[PR-104]` please?

File: src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java [hunk 2]
```diff
@@ -112,13 +114,10 @@
...
-    public static final String FILE_SEPARATOR_SUBST = "_";
-    /**
-     * Prefix for generated sources stored as a separate artifact in cache
-     */
-    private static final String BUILD_PREFIX = "build" + FILE_SEPARATOR_SUBST;
-
     private static final Logger LOGGER = LoggerFactory.getLogger(CacheControllerImpl.class);
+    private static final String DEFAULT_FILE_GLOB = "*";
+    public static final String ERROR_MSG_RESTORATION_OUTSIDE_PROJECT =
+            "Blocked an attempt to restore files outside of a project directory : ";
```
Comment (hboutemy) [inline src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java:120]: in english, no space before `:` (yes, french with non-breaking space is much better, but we must adapt :) )

File: src/main/java/org/apache/maven/buildcache/artifact/OutputType.java (hunk 1 above)
Comment (hboutemy) [reply src/main/java/org/apache/maven/buildcache/artifact/OutputType.java:23]: yes, ARTIFACT in english, not nice french ARTEFACT :)

File: src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java [hunk 3]
```diff
@@ -710,6 +711,18 @@ public static boolean isRestoreGeneratedSources(MavenProject project) {
                 project.getProperties().getProperty(RESTORE_GENERATED_SOURCES_PROPERTY_NAME, "true"));
     }
 
+    /**
+     * Allow skipping artefacts restoration on a per-project level via a property (which defaults to true)
```
Comment (hboutemy) [inline src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java:715]: s/artefacts/artifacts/ (in comment, not critical)

File: src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java [hunk 4]
```diff
@@ -133,6 +132,15 @@ public class CacheControllerImpl implements CacheController {
     private final RestoredArtifactHandler restoreArtifactHandler;
     private volatile Scm scm;
 
+    /**
+     * A map dedicated to store the base path of resources stored to the cache which are not original artefacts
```
Comment (hboutemy) [inline src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java:136]: s/artefacts/artifacts/

File: src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java [hunk 5]
```diff
@@ -92,16 +92,17 @@
 import static org.apache.maven.buildcache.xml.CacheConfigImpl.CACHE_ENABLED_PROPERTY_NAME;
 import static org.apache.maven.buildcache.xml.CacheConfigImpl.CACHE_SKIP;
 import static org.apache.maven.buildcache.xml.CacheConfigImpl.RESTORE_GENERATED_SOURCES_PROPERTY_NAME;
+import static org.apache.maven.buildcache.xml.CacheConfigImpl.RESTORE_ON_DISK_ARTEFACTS_PROPERTY_NAME;
```
Comment (hboutemy) [inline src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java:95]: s/ARTEFACT/ARTIFACT/

File: src/main/java/org/apache/maven/buildcache/xml/CacheConfig.java [hunk 6]
```diff
@@ -131,5 +132,10 @@ public interface CacheConfig {
      */
     boolean isRestoreGeneratedSources();
 
+    /**
+     * Flag to restore (default) or not generated artefacts
+     */
+    boolean isRestoreOnDiskArtefacts();
```
Comment (hboutemy) [inline src/main/java/org/apache/maven/buildcache/xml/CacheConfig.java:138]: s/artefacts/artifacts/

File: src/main/java/org/apache/maven/buildcache/xml/CacheConfigImpl.java [hunk 7]
```diff
@@ -90,6 +91,7 @@ public class CacheConfigImpl implements org.apache.maven.buildcache.xml.CacheCon
     public static final String FAIL_FAST_PROPERTY_NAME = "maven.build.cache.failFast";
     public static final String BASELINE_BUILD_URL_PROPERTY_NAME = "maven.build.cache.baselineUrl";
     public static final String LAZY_RESTORE_PROPERTY_NAME = "maven.build.cache.lazyRestore";
+    public static final String RESTORE_ON_DISK_ARTEFACTS_PROPERTY_NAME = "maven.build.cache.restoreOnDiskArtefacts";
```
Comment (hboutemy) [inline src/main/java/org/apache/maven/buildcache/xml/CacheConfigImpl.java:94]: same

**Review by hboutemy — CHANGES_REQUESTED** [review]
you should just search global "artefact" and fix :)

the new configuration property should probably be documented in https://maven.apache.org/extensions/maven-build-cache-extension/parameters.html

and as written before, please change the PR-104 in the commit message

**Review by AlexanderAshitkin — APPROVED** [review]

**Comment by AlexanderAshitkin** [comment]
Hi, @kbuntrock @hboutemy. This PR looks good. How can we move it from a draft to an open state?

**Comment by olamy** [comment]
> you should just search global "artefact" and fix :)
> 

Both are correct it only depends if we consider speaking English from UK or English from US
But hey original English is from UK :)

**Comment by kbuntrock** [comment]
> Hi, @kbuntrock @hboutemy. This PR looks good. How can we move it from a draft to an open state?

Hello @AlexanderAshitkin. I guess this PR needs now a bit of "huile de coude" like we say in France ("elbow oil" ...). Meaning that I have to work. 😊

In details, I need to : 
- take into account your reviews
- update the documentation
- add some tests

File: src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java (hunk 2 above)
Comment (kbuntrock) [reply src/main/java/org/apache/maven/buildcache/CacheControllerImpl.java:120]: Gosh, I should know, it is precisely the reason how I got a Apache Jira account (and my very first contribution to the open source world 😋) : https://issues.apache.org/jira/browse/CXF-7185

**Comment by kbuntrock** [comment]
The code introduced with MBUILDCACHE-80 has partially the same purpose of this MR. I need to adapt my code.
First version of the merge can be seen in commit : https://github.com/apache/maven-build-cache-extension/pull/104/commits/c2878e17514ac909343e0ff10898b3bcbab0c2a7

But I still have some work since restored `target` directory shows some "weird" pom files.

image

**Comment by hboutemy** [comment]
> some "weird" pom files

oh, this is Maven 4 consumer POMs, that seem to have been generated with random file names

@gnodet do we really want these random file names? Do we need to create a hack in build cache extension to manage this randomness?

**Comment by kbuntrock** [comment]
> > some "weird" pom files
> 
> oh, this is Maven 4 consumer POMs, that seem to have been generated with random file names
> 
> @gnodet do we really want these random file names? Do we need to create a hack in build cache extension to manage this randomness?

To add a bit of context, I did some experiment yesterday evening. Here are some extract of the file `buildinfo.xml` after the execution of the goal package in different contexts: 

### Test "IncrementalRestoreTest", on this branch:

```xml
<goals>
  <goal>package</goal>
</goals>
<artifact>
  <groupId>org.apache.maven.caching.test</groupId>
  <artifactId>mbuildcache-incremental</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <type>jar</type>
  <fileName>mbuildcache-incremental.jar</fileName>
  <fileHash>887c667918b9b71d</fileHash>
  <fileSize>3119</fileSize>
  <filePath>target/mbuildcache-incremental-final.jar</filePath>
</artifact>
<attachedArtifacts>
  <attachedArtifact>
    <groupId>org.apache.maven.caching.test</groupId>
    <artifactId>mbuildcache-incremental</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <classifier>consumer</classifier>
    <type>pom</type>
    <fileName>mbuildcache-incremental-consumer.pom</fileName>
    <fileHash>bf52cc397806673a</fileHash>
    <fileSize>430</fileSize>
    <filePath>target/consumer-4676390733155918308.pom</filePath>
  </attachedArtifact>
</attachedArtifacts>
```

### Test "IncrementalRestoreTest", on the branch master:

```xml
<goals>
  <goal>package</goal>
</goals>
<artifact>
  <groupId>org.apache.maven.caching.test</groupId>
  <artifactId>mbuildcache-incremental</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <type>jar</type>
  <fileName>mbuildcache-incremental.jar</fileName>
  <fileHash>41e8d89e0385f771</fileHash>
  <fileSize>3119</fileSize>
</artifact>
<attachedArtifacts>
  <attachedArtifact>
    <groupId>org.apache.maven.caching.test</groupId>
    <artifactId>mbuildcache-incremental</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <classifier>consumer</classifier>
    <type>pom</type>
    <fileName>mbuildcache-incremental-consumer.pom</fileName>
    <fileHash>bf52cc397806673a</fileHash>
    <fileSize>430</fileSize>
  </attachedArtifact>
</attachedArtifacts>
```


### On a standalone project based on "IncrementalRestoreTest", with the extension code of this branch:

```xml
<goals>
  <goal>package</goal>
</goals>
<artifact>
  <groupId>org.apache.maven.caching.test</groupId>
  <artifactId>mbuildcache-incremental</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <type>jar</type>
  <fileName>mbuildcache-incremental.jar</fileName>
  <fileHash>8d6a9a9795c1f249</fileHash>
  <fileSize>3122</fileSize>
  <filePath>target/mbuildcache-incremental-final.jar</filePath>
</artifact>
```

Meaning that:
- It might be linked to the IT execution context
- It is not related to this PR, so I will focus on updating the current tests and put this problem aside.

**Comment by hboutemy** [comment]
notice: if you build with Maven 3, you won't have this extra file, which is Maven 4 specific https://cwiki.apache.org/confluence/display/MAVEN/Build+vs+Consumer+POM

**Comment by kbuntrock** [comment]
Alright, I think I've completed my TODO list. :) (@hboutemy & @AlexanderAshitkin)

**Comment by kbuntrock** [comment]
Alright, javadoc generation should be fixed

**Review by hboutemy — APPROVED** [review]
LGTM

**Comment by kbuntrock** [comment]
Just for clarity, #103 should be merged first.

**Comment by kbuntrock** [comment]
> @kbuntrock #103 has been merged. Thanks!

@olamy: rebased, ready to merge :)

**Comment by martincekodhima** [comment]
Thanks everyone! 😄

**Comment by nobesio** [comment]
Thank you, @kbuntrock!

**Comment by kbuntrock** [comment]
> Thank you, @kbuntrock!

You're welcome :)

**Comment by hacosta** [comment]
I'm really looking forward to testing this out, is there a new release planned or a place where I can track it?

**Comment by kbuntrock** [comment]
> I'm really looking forward to testing this out, is there a new release planned or a place where I can track it?

Yes, Olivier mentioned on the maven dev list that he is planning to create a release probably this week. (https://lists.apache.org/list.html?dev@maven.apache.org)

**Comment by olamy** [comment]
vote started https://lists.apache.org/thread/do2dc2tpmq2zknyqnf55xpvllop94bq5

if not issues this should be available early next week

**Comment by julien-pcd** [comment]
The change to "Forbid the possibility to extract/restore data in a directory outside the project" is a breaking change for some of my projects. In my case, I have several multi-module projects with child POMs configured to use ../target/${project.artifactId} as their build directory in order to have a single target dir in the parent rather than N children target directories.

**Comment by olamy** [comment]
I would say option 1. 
Option 2 might to complicated to implement?

**Comment by jira-importer** [comment]
Resolve #318

**Pull request merged**